#2 parse `FriendsOfPhp/security-advisories` example YAML into sensible data structures (that aren't terrible YAML)

Ocramius opened the pull request:
Fixes #2

Ocramius committed:
Committing files as provided by `stack init`

Ocramius committed:
Imported project sources as per project initial structure by `stack new`

Ocramius committed:
Adding license, required by `stack test`

Ocramius committed:
Added `package.yaml` (had to figure out that `hpack` is a thing O_o) and added `yaml` dependency

Ocramius committed:
Moved YAML dependency into `library`, since it isn't really relevant for the package itself

Ocramius committed:
First successful compile + test run according to Hspec

Parsed this thing into something typed, that makes sense and ISN'T SHITTY YAML!

Ocramius commented on LICENSE:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2014 Roave, LLC
```
Yes, I'm planning for this to live under `Roave/` once done.

Ocramius commented on SecurityAdvisoriesBuilder-hs.cabal:
```diff
@@ -0,0 +1,60 @@
+-- This file has been generated from package.yaml by hpack version 0.28.2.
```
Given that `hpack` is used to generate this, should this be under `.gitignore`?

Akii replied:
Yes. Last time I bootstrapped a package with stack it actually created a .gitignore with the cabal file excluded. O.o

Ocramius commented on app/Main.hs:
```diff
@@ -0,0 +1,6 @@
+module Main where
```
This comes from `stack new`

Akii replied:
adjusted the garbage

Ocramius commented on src/Data/String/Strip.hs:
```diff
@@ -0,0 +1,6 @@
+module Data.String.Strip (strip)  where
```
This comes from `stack new`

Akii replied:
removed the garbage

Ocramius commented on src/Roave/SecurityAdvisories/FriendsOfPhp.hs:
```diff
@@ -0,0 +1,45 @@
+
+data AffectedBranch = AffectedBranch {
+  time :: Maybe Text
+  , versions :: [Text]
```
`[Text]` is obviously too simplistic, but it will do until we decide to go further into detail

Ocramius replied:
From what I read, I will need to implement my own `FromJSON` instance once I defined the specification for a `VersionBoundary`

Akii replied:
Yes. Good for now (except for the formatting of course :P)

Ocramius commented on package.yaml:
```diff
@@ -0,0 +1,35 @@
+library:
+  source-dirs: src/
+  dependencies:
+    - bytestring >= 0.10.8.2
```
Note: instead of putting the dependencies at root level, I specified them separately for `library` and `tests` - is that a good or a bad idea? Any pointers to how real world projects do this?

Akii replied:
Yes, that's common practice. Wire is always a good source of real world projects: https://github.com/wireapp/wire-server/blob/develop/services/galley/galley.cabal

Ocramius commented on src/Roave/SecurityAdvisories/FriendsOfPhp.hs:
```diff
@@ -0,0 +1,45 @@
+    v .: "cve" <*>
+    v .: "branches" <*>
+    v .: "reference"
+  parseJSON _ = fail "Expected an Object to be parsed into an Advisory"
```
Not a fan of this kind of simplistic failure, but it works for now

Ocramius commented on src/Roave/SecurityAdvisories/FriendsOfPhp.hs:
```diff
@@ -0,0 +1,45 @@
+    AffectedBranch <$>
+    v .: "time" <*>
+    v .: "versions"
+  parseJSON _ = fail "Expected an Object to be parsed into an AffectedBranch"
```
Not a fan of this kind of simplistic failure, but it works for now

Akii replied:
What would you rather have?

Ocramius replied:
Generally not a `String`, but it is indeed the exception case (which I need to test)

Ocramius commented on src/Roave/SecurityAdvisories/FriendsOfPhp.hs:
```diff
@@ -0,0 +1,45 @@
+instance FromJSON AffectedBranch where
+  parseJSON (Y.Object v) =
+    AffectedBranch <$>
+    v .: "time" <*>
```
Note: I don't really understand this syntax yet - will likely need to discuss it in a call, when possible.

Akii replied:
Can do. Need to go through that myself as well.

Ocramius commented on stack.yaml:
```diff
@@ -0,0 +1,71 @@
+#
+# resolver: ./custom-snapshot.yaml
+# resolver: https://example.com/snapshots/2018-01-01.yaml
+resolver: lts-12.9
```
The resolver is quite obscure to me: if I understand this correctly, it is just a repository version. What I did is searching for packages on hackage, then pick the latest version (as one should) and finally adding them here under `extra-deps`. Is that OK or is the preferred approach another one?

Akii replied:
A resolver refers to a snapshot (I think there is an ongoing discussion whether this should be renamed). A snapshot is a set of packages pinned at a specific version. A snapshot also has a specific GHC version (in our case 8.4.3).

So, knowing that, it is common (I'd dare say that's the whole point), that a snapshot does not have the latest version of packages. It has a set of packages that are known to work (= are stable). They undergo testing before being accepted into a LTS snapshot.

Now, "picking the latest version" isn't what I've been doing in Haskell. In my experience, unless you need a specific feature that has recently been added, sticking to the package versions provided by the snapshot is fine/encouraged.

When a package is not available in a snapshot, or at a lower version, you can add it to `extra-deps` essentially modifying the snapshot (does that make sense? like an override). And this should be the only reason you ever add a package there.

You can check out which packages are available and in which versions by looking at the snapshot on stackage:

https://www.stackage.org/lts-12.9
https://www.stackage.org/lts-12.9/package/yaml-0.8.32

So instead of going to Hackage and taking the newest version, I'd suggest you go to Stackage. I've asked Twitter if it is even necessary to add version constraints. I assume it is, but never do it for my personal projects.

</wall-of-text>

Did a bit of testing and turns out all you need to add is:

```yaml
extra-deps:
- yaml-0.10.1.1
```

as well as (in the package.yaml dependencies)

```yaml
- containers >= 0.5.11.0
```

and it compiles.

Ocramius replied:
All clear. One question remains though: is the resolver immutable? If it is mutable (updates, flow in), then using version constraints is probably wise anyway

Akii replied:
Good question, I had no idea. Apparently major versions are locked. Minor versions (and, therefore, hopefully without breaking API changes) get updates: https://github.com/commercialhaskell/stackage#frequently-asked-questions

TIL

Ocramius replied:
As discussed on the twitters, I will use the caret operator for all major versions (where `0.x.y.z` has `x` as major in hackage/stackage)

Ocramius commented on test/Data/String/StripSpec.hs:
```diff
@@ -0,0 +1,19 @@
+module Data.String.StripSpec (main, spec) where
```
added by `stack new`

Akii replied:
No idea what template you used but it generated a lot of crap :P I'll remove the garbage and push.

Ocramius replied:
No template at all: just `stack new` as per stack 1.7.1

Ocramius commented on test/Roave/SecurityAdvisories/FriendsOfPhpSpec.hs:
Not sure how to annotate scenarios correctly.

Akii replied:
The defacto standard is haddock, which is similar to phpdoc comments. 
https://haskell-haddock.readthedocs.io/en/latest/

Ocramius commented on test/Roave/SecurityAdvisories/FriendsOfPhpSpec.hs:
```diff
@@ -0,0 +1,46 @@
+main = hspec spec
+
+exampleAdvisory :: ByteString
+exampleAdvisory = [r|
```
I was quite confused by `Text.RawString.QQ.r` and `QuasiQuotes` - is this the normal way to define multi-line strings?

Akii replied:
It can actually do much more than that. For example, there is a HTML templating engine using QuasiQuotes named Shakespeare (https://www.yesodweb.com/book/shakespearean-templates).

Ocramius commented on test/Roave/SecurityAdvisories/FriendsOfPhpSpec.hs:
I was wondering if there was a decent way to pass this through quickcheck, instead of manually providing examples?

Akii replied:
Let's use hedgehog: http://hackage.haskell.org/package/hedgehog

Ocramius replied:
Will create an issue for this

Ocramius commented on src/Roave/SecurityAdvisories/FriendsOfPhp.hs:
```diff
@@ -0,0 +1,45 @@
+} deriving (Eq, Show)
+
+parseAdvisoryYaml :: ByteString -> Either String Advisory
+parseAdvisoryYaml = Y.decodeEither
```
Wanted to use `decodeEither'`, but it doesn't provide an instance of `Eq`, so it cannot be used while parsing. `decodeEither` is deprecated. Any ideas welcome :+1:

Akii replied:
I'm getting this type error:

```
    • Couldn't match type ‘ParseException’ with ‘[Char]’
      Expected type: ByteString -> Either String Advisory
        Actual type: ByteString -> Either ParseException Advisory
    • In the expression: decodeEither'
      In an equation for ‘parseAdvisoryYaml’:
          parseAdvisoryYaml = decodeEither'
   |
29 | parseAdvisoryYaml = decodeEither'
   |                     ^^^^^^^^^^^^^
```

Ocramius replied:
Yes, that's because I added the tyoe declaration on `parseAdvisoryYaml`: if you adapt it, it is not possible to use HSpec's `shouldBe` with this type

Akii replied:
Gotcha, let's delay the decision until we know what kind of error we need.

Akii replied:
You can use the BiFunctor instance for Either to map the left side. What you'll need is a morphism from `ParseException` to `String` (probably want to use Text here).

https://hackage.haskell.org/package/bifunctors-5/docs/Data-Bifunctor.html

Ocramius replied:
If I understood this correctly, what you are suggesting is to have an `instance Bifunctor ParseException where ...` in order for the test to remain unchanged?

Akii committed:
Remove garbage generated by stack new and do some formatting with hlint/stylish-haskell

Akii commented:
Ah yes, one more thing: Use hindent and stylish-haskell. They indent code in a standardised / good way.

Akii commented on package.yaml:
```diff
@@ -0,0 +1,35 @@
+library:
+  source-dirs: src/
+  dependencies:
+    - bytestring >= 0.10.8.2
```
You might want to a) not specify the version or b) only specify the upper bound (see 
https://twitter.com/ProgrammerDude/status/1041239338795851777).
Reason being that you use stack which already pins package versions by the resolver (snapshot).

Akii commented:
Since you'll be using a lot of `OverloadedStrings` I'd suggest adding it as default extension:

```yaml
library:
  default-extensions:
  - OverloadedStrings
```

Ocramius committed:
Defining default extensions rather than enabling extensions in each file

As specified by @Akii in https://github.com/Ocramius/SecurityAdvisoriesBuilder-hs/pull/3#issuecomment-421723368

Ocramius committed:
Removed `.cabal` definitions from git index, since `hpack` generates those files for us

Ref: @Akii at https://github.com/Ocramius/SecurityAdvisoriesBuilder-hs/pull/3#discussion_r217897751

Ocramius commented:
> Ah yes, one more thing: Use hindent and stylish-haskell. They indent code in a standardised / good way.

Hmm, I had those configured and set up correctly in the IDE, but apparently it's not running them as expected.

Ocramius committed:
As discussed on twitter, dropping overly-new dependencies and using current solver latest versions

This also means sticking to `>=` + `<` dependency constraints, since the caret
operator is only introduced in Cabal 2.0+, which isn't enforced by hpack.

Ref: https://twitter.com/ProgrammerDude/status/1041428729954791424
Ref: https://www.haskell.org/cabal/users-guide/developing-packages.html#pkg-field-build-depends

Ocramius committed:
Added travis-ci configuration

Ocramius commented:
```
The command "stack --no-terminal test" exited with 0.
```

Aaaaand merge :+1:
